[knx] Add support for KNX IP Secure

holgerfriedrich opened the pull request:
* add support for KNX IP Secure, new options SECURETUNNEL and
  SECUREROUTER, refers to #8872
* add config options for credentials for secure connections
* update user documentation
* add test cases

Signed-off-by: Holger Friedrich <mail@holger-friedrich.de>

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").

ATTENTION: Don't use "git merge" when working with your pull request branch!
This can clutter your Git history and make your PR unuseable.
Use "git rebase" instead. See this forum post for further details:
https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

Add one or more appropriate labels to make your PR show up in the release notes.
E.g. enhancement, bug, documentation, new binding
This can only be done by yourself if you already contributed to this repo.

If your PR's code is not backward compatible with previous releases (which
should be avoided), add a message to the release notes by filing another PR:
https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst

# Title

Provide a short summary in the *Title* above. It will show up in the release notes.
For example:
- [homematic] Improve communication with weak signal devices
- [timemachine][WIP] Initial contribution

# Description

Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work

# Testing

Your pull request will automatically be built and available under the following folder:
https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/

It is a good practice to add a URL to your built JAR in this pull request description,
so it is easier for the community to test your Add-on.
If your pull request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Marketplace:
https://community.openhab.org/c/marketplace/69

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

holgerfriedrich commented:
This is basically the next microstep towards supporting KNX Secure.
I just separated out the functions needed for the secure connection between OH and a Secure IP Interface or a Secure Router.
In contrast to #12434, there is no keyring file yet, the passwords need to be provided manually.
This keeps the PR smaller for now and avoids the situation that we need to explain why we have KNX Data Secure only partially supported (read only).

@slueder I would appreciate if you can test this one as well on your hardware.

@kaikreuzer how shall we proceed with the support for Secure Router? Until now we found no one willing to test. We both have only Secure IP interfaces.

kaikreuzer commented:
> @kaikreuzer how shall we proceed with the support for Secure Router? Until now we found no one willing to test. We both have only Secure IP interfaces.

I'd say that is up to you to decide. If you believe that it should work in general, keep the support in the code. If it is rather unlikely that it'll work untested, better remove support for it. Up to you to judge!

holgerfriedrich commented:
@kaikreuzer any chance to push this forward in the next few days? I would appreciate it if this could make it into M7.

holgerfriedrich commented:
@lolodomo do you want to have a look at this as well? I know it is up to Kai to approve this, but maybe you can already give some recommendations....

lolodomo commented:
> @lolodomo do you want to have a look at this as well? I know it is up to Kai to approve this, but maybe you can already give some recommendations....

I let Kai review and merge.

holgerfriedrich commented:
@kaikreuzer how can we proceed with this? The patch is there for quite a while now and I would appreciate if it could be merged. For me it works fine and stable on OH 3.3 now (IP interface, tunnel mode).

florian-h05 commented:
@kaikreuzer Gentle ping to not forget this PR.

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -70,20 +110,64 @@ public void initialize() {
+            boolean tunnelOk = ((secureTunnel.user > 0) && (secureTunnel.devKey.length == 16)
+                    && (secureTunnel.userKey.length == 16));
+            if (!tunnelOk) {
+                logger.warn("Bridge {} incomplete security configuration for secure tunnel", thing.getUID());
```
Could you add specific checks for the parameters, so that you could directly hint to the user, which parameter he forgot to provide?

holgerfriedrich replied:
Not sure if I should add individual checks.... This is anyway an advanced setup and requires reading the manual.
The next PR for data sec will introduce a second way to configure the items;: the keyring.
If you are fine with it, I would leave it as it is for now.

kaikreuzer replied:
Ok for me.

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -70,20 +110,64 @@ public void initialize() {
+                return;
+            }
+
+            logger.info("KNX secure tunneling needs a few seconds to establish connection");
```
reduce to debug

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -70,20 +110,64 @@ public void initialize() {
+            ipConnectionType = IPClient.IpConnectionType.SECURE_ROUTER;
+
+            if (!securityAvailable) {
+                logger.warn("Bridge {} missing security configuration for secure routing", thing.getUID());
```
Could you add specific checks for the parameters, so that you could directly hint to the user, which parameter he forgot to provide?

holgerfriedrich replied:
check for is directly below for secure router.
this if is just for the case when security is not configured at all....

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -70,20 +110,64 @@ public void initialize() {
+            }
+            if (secureRouting.backboneGroupKey.length != 16) {
+                // failed to read shared backbone group key from config
+                logger.warn("Bridge {} missing security configuration for secure routing", thing.getUID());
```
Could you add specific checks for the parameters, so that you could directly hint to the user, which parameter he forgot to provide?

holgerfriedrich replied:
this is exactly the check for secure routing. only one parameter....

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -70,20 +110,64 @@ public void initialize() {
+                        "backboneGroupKey required for secure routing; please configure");
+                return;
+            }
+            logger.info("KNX secure routing needs a few seconds to establish connection");
```
reduce to debug

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -70,20 +110,64 @@ public void initialize() {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                    MessageFormat.format("Unknown IP connection type {0}. Known types are either 'TUNNEL' or 'ROUTER'",
-                            connectionTypeString));
+            logger.warn("Bridge {} unknown connection type", thing.getUID());
```
reduce to debug

holgerfriedrich replied:
ok, I considered setting the level to debug as hiding the error. But I am fine with reducing to debug. events.log will at complain about this as well on INFO level.

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -95,23 +179,42 @@ public void initialize() {
+            initJob = null;
         }
+
+        {
```
These brackets can be removed.

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/KNXBridgeBaseThingHandler.java:
```diff
@@ -29,27 +29,135 @@
+
+        // step 1: secure routing, backbone group key manually specified in OH config
+        if (!cRouterBackboneGroupKey.isBlank()) {
+            // provided in conig
```
typo: config

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/KNXBridgeBaseThingHandler.java:
```diff
@@ -29,27 +29,135 @@
+        // step 1: secure routing, backbone group key manually specified in OH config
+        if (!cRouterBackboneGroupKey.isBlank()) {
+            // provided in conig
+            String tmp = cRouterBackboneGroupKey.trim().replaceFirst("^0x", "").trim().replace(" ", "");
```
Is there a better name for the variable than `tmp`?

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/resources/OH-INF/thing/ip.xml:
```diff
@@ -64,6 +72,32 @@
 				<description>Seconds between connection retries when KNX link has been lost, 0 means never retry, minimum 30s</description>
 				<default>60</default>
 			</parameter>
+			<parameter name="routerBackboneKey" type="text" groupName="knxsecure">
```
Are those parameters hidden when the group is hidden? Or is it necessary to also mark them as `advanced`? It would imho be cleaner.

holgerfriedrich replied:
Good point. Actually, it did not properly work to set "advanced" on group level.
I moved "advanced" down to the elements and added a note to the SECURE settings `<option value="SECURETUNNEL">Secure tunnel (experimental, use advanced options to configure)</option>`
Cleaner.

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/resources/OH-INF/thing/ip.xml:
```diff
@@ -64,6 +72,32 @@
+				<description>KNX secure, optional: Backbone key for secure router mode. 16 bytes
+					in hex notation. Can also be found
+					in ETS security report.</description>
+				<default></default>
```
Please remove empty defaults.

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/resources/OH-INF/thing/ip.xml:
```diff
@@ -64,6 +72,32 @@
+			</parameter>
+			<parameter name="tunnelDeviceAuthentication" type="text" groupName="knxsecure">
+				<context>password</context>
+				<label>
```
remove the linebreak

kaikreuzer commented on bundles/org.openhab.binding.knx/src/main/resources/OH-INF/thing/ip.xml:
```diff
@@ -64,6 +72,32 @@
+			<parameter name="routerBackboneKey" type="text" groupName="knxsecure">
+				<context>password</context>
+				<label>Router backbone key</label>
+				<description>KNX secure, optional: Backbone key for secure router mode. 16 bytes
```
Please remove the "KNX secure, optional:" from all parameters - they anyhow appear in the "KNX secure" parameter group.

kaikreuzer reviewed (changes requested):
Hi @holgerfriedrich,
Very sorry for having kept you waiting for so long!
I've now reviewed the PR and left a few comments on it.
Should be easy to address, so that all should be fine to be merged very soon!

holgerfriedrich commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/client/AbstractKNXClient.java:
```diff
@@ -146,15 +157,19 @@ public void initialize() {
+            // schedule connect job, for the first connection ignore autoReconnectPeriod and use 1 sec
+            final long reconnectDelayS = (state == ClientState.INIT) ? 1 : autoReconnectPeriod;
+            final String prefix = (state == ClientState.INIT) ? "re" : "";
+            logger.info("Bridge {} scheduling {}connect in {}s", thingUID, prefix, reconnectDelayS);
```
done

holgerfriedrich commented on bundles/org.openhab.binding.knx/README.md:
```diff
@@ -208,6 +212,30 @@ Each configuration parameter has a `mainGA` where commands are written to and op
+
+### KNX IP Secure
+
+KNX IP Secure protects the traffic between openHAB and your KNX installation. It **requires a KNX Secure Router or a Secure IP Interface** and a KNX installation **with security features enabled in ETS tool**.
```
done

holgerfriedrich commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -56,7 +59,44 @@ public IPBridgeThingHandler(Bridge bridge, NetworkAddressService networkAddressS
+        initJob = scheduler.submit(() -> {
+            initializeLater();
+        });
+    }
```
would `updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.CONFIGURATION_PENDING, "initialization scheduled");` be appropriate?

jlaur replied:
You should omit the last two arguments - see https://www.openhab.org/docs/concepts/things.html#status-details

holgerfriedrich commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/client/IPClient.java:
```diff
@@ -107,9 +159,27 @@ private KNXnetIPConnection getConnection(int serviceMode, @Nullable InetSocketAd
+
+                    logger.trace("creating new TCP connection");
+                    if (tcpSession != null) {
+                        logger.warn("tcpSession might still be open");
```
I consider this as a bug situation. I ran into the issue that too many connections were created and OH could no longer connect to the IP interface. This took some efforts to be resolved, and it seems fine now.
However, I want to see bug reports if it happens on other hardware. -> add text like "please report this issue on GitHub"?

kaikreuzer replied:
Yes, please add something that encourages the user to report the issue.

holgerfriedrich replied:
Same as below. Switching to debug. If anyone experiences problems with connection overflow again, debug logging is needed anyway.

holgerfriedrich commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -70,20 +110,64 @@ public void initialize() {
+            ipConnectionType = IPClient.IpConnectionType.SECURE_TUNNEL;
+
+            if (!securityAvailable) {
+                logger.warn("Bridge {} missing security configuration for secure tunnel", thing.getUID());
```
this is the "no config at all case"

holgerfriedrich commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/client/IPClient.java:
```diff
@@ -119,11 +189,39 @@ private KNXnetIPConnection getConnection(int serviceMode, @Nullable InetSocketAd
+                } catch (InterruptedException e) {
+                }
+            } catch (Exception e) {
+                logger.warn("closing TCP connection failed: {}", e.getMessage());
```
Not sure about failure cases during close. Calimero cancels some futures, sends connection close packets, etc.
I have the feeling this could happen due to runtime issues. Have not seen it recently, but I don't want to hide it.
-> will change to info

kaikreuzer replied:
You should then reduce it to debug. What is a user supposed to do when he sees that info message? It mainly seems relevant for you and you'll likely be running in debug anyhow. 😉

holgerfriedrich replied:
fair enough. debug.

holgerfriedrich commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/client/AbstractKNXClient.java:
```diff
@@ -171,78 +186,139 @@ private synchronized boolean connectIfNotAutomatic() {
             scheduleReconnectJob();
             return false;
+        } catch (KNXIllegalArgumentException e) {
+            logger.warn("Bridge {} cannot connect: {}", thingUID, e.getMessage());
```
`disconnect(e)` will set the ThingStatus accordingly. Setting to debug.

holgerfriedrich commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/client/AbstractKNXClient.java:
```diff
@@ -171,78 +186,139 @@ private synchronized boolean connectIfNotAutomatic() {
+
+            return false;
+        } catch (KNXException | KnxSecureException e) {
+            logger.warn("Bridge {} cannot connect: {}", thingUID, e.getMessage());
```
`disconnect(e)` will set the ThingStatus accordingly. Setting to debug.

jlaur commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -56,7 +59,45 @@ public IPBridgeThingHandler(Bridge bridge, NetworkAddressService networkAddressS
+            } else {
+                logger.debug("KNX security not configured");
+            }
+        } catch (NullPointerException e) {
```
You should not catch `NullPointerException`. If this is a bug within the binding, it should be fixed. If it's a bug within an external library outside our control, you can surround API call by try/catch and catch `Exception`.

holgerfriedrich replied:
I know this looks a bit strange, but actually this is some kind of graceful degrading.
The extended version of this PR for data secure contains reading a KNX keyring file within `initializeSecurity()` (using Calimero library). I think I have seen NPEs out of my control in the past, this is why I added this catch block. 
The effect is basically that the addon is running without security support instead of crashing.

But I think it is also fine to remove this. WDYT?

jlaur replied:
If `initializeSecurity()` makes an API call to Calimero, I would propose to explicitly guard that call:
```
try {
    callCalimero();
}
catch (Exception e) {
    // This is a bug, handle gracefully - just as example:
    throw KnxException("Unexpected Calimero exception", e)
}
```

Here and below I would then explicitly catch checked exceptions from own code and documented runtime exceptions from Calimero if they are not mapped/rethrown from own code.

In other words: Yes, it would be fine to remove it, if at the same time making sure we have guards elsewhere to make sure the binding works reliably even if Calimero misbehaves.

holgerfriedrich replied:
Ok, will use this pattern for the extended PR to guard the call to the base library. This is anyway not part of this PR.

`initializeSecurity()` now only throws `KNXSecureException`. I checked the function calls within `initializeSecurity` again and currently everything seems to be covered by catching `KNXSecureException`.

jlaur replied:
Small correction to example - you probably get the point, but now reading it again, it must be corrected. :)

```
try {
    callCalimero();
}
catch (KnxException e) { // and others expected
    throw e; // if not handled here
}
catch (Exception e) {
    // This is a bug, handle gracefully - just as example:
    logger.trace("", e);
    throw KnxException("Unexpected Calimero exception", e)
}
```

jlaur commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/KNXBridgeBaseThingHandler.java:
```diff
@@ -29,27 +29,135 @@
+                parsed[i] = (byte) Integer.parseInt(hexstring.substring(2 * i, 2 * i + 2), 16);
+            }
+        } catch (NumberFormatException e) {
+            throw new KnxSecureException("backbone key configured, cannot parse hex string, illegal character");
```
```suggestion
            throw new KnxSecureException("backbone key configured, cannot parse hex string, illegal character", e);
```

jlaur commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/KNXBridgeBaseThingHandler.java:
```diff
@@ -29,27 +29,135 @@
+     * @param hexstring 32 characters hex
+     * @return key in byte array format
+     */
+    public static byte[] secHelperParseBackboneKey(String hexstring) throws NumberFormatException, KnxSecureException {
```
```suggestion
    public static byte[] secHelperParseBackboneKey(String hexstring) throws KnxSecureException {
```

jlaur commented on bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/handler/IPBridgeThingHandler.java:
```diff
@@ -56,7 +59,45 @@ public IPBridgeThingHandler(Bridge bridge, NetworkAddressService networkAddressS
+            logger.warn("Bridge {}, initialization failed", thing.getUID(), e); // for NPE log a stack trace as well
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "KNX security: NPE, see log");
+            return;
+        } catch (Exception e) {
```
So here I would also change it to catch the KNX exceptions explicitly.

jlaur commented:
@holgerfriedrich - please consider my comments only as additional "fly by" comments, and I hope they do not interfere with the review @kaikreuzer is currently conducting.

A general tip, though: I would suggest to push new commits, as the review progresses, without rebasing. This makes it possible to track the changes, and we can easily clean up the comments when performing a squash-merge eventually. The current approach theoretically requires a full review after each force-push as it's not possible to see what changed.

holgerfriedrich commented:
@jlaur thanks for your review - and also for the note about the force pushes I am using! I was actually unsure if I should stash my changes. I did not rebase recently, but as this PR was around for a while I had to do so at some point in time.

@kaikreuzer Thanks for review, hopefully I addressed all comments.

kaikreuzer reviewed (approved):
lgtm, many thanks!
I leave the PR to @jlaur for merging then.
